### ISSUE TITLE
Error when substituting into empty polynomial

### ISSUE THREAD

**Issue by rdeits** [blegat/MultivariatePolynomials.jl#28]
```julia
julia> using MultivariatePolynomials

julia> @polyvar x y

julia> z = differentiate(x + 1, y)

julia> subs(z, [0], [x])
ERROR: ArgumentError: reducing over an empty collection is not allowed
 in _mapreduce(::MultivariatePolynomials.##98#99{MultivariatePolynomials.Polynomial{true,Int64},Array{Int64,1}}, ::Base.#+, ::Base.LinearFast, ::UnitRange{Int64}) at ./reduce.jl:154
 in subs(::MultivariatePolynomials.Polynomial{true,Int64}, ::Array{Int64,1}, ::Array{MultivariatePolynomials.PolyVar{true},1}) at /Users/rdeits/.julia/v0.5/MultivariatePolynomials/src/subs.jl:77
```

I'll see if I can fix this one 😄

**Comment by rdeits** [comment]
There's a trivial fix for this, which is to add `length(p) == 0 && return 0` to the body of `subs()`. But that introduces an obvious type-instability, which in turn raises a deeper question: what *should* the return type of `subs()` be? Currently, it's ambiguous: if the user provides a full set of variable assignments, then the return type is numeric, but if the user does not, then the return type is a Polynomial over the remaining variables. There's no way for inference to know which case will apply to a given polynomial, so type instability is unavoidable. 

We could instead decide to always return a Polynomial and force the user to convert it to a numeric type if desired. That would be type-stable but possibly less efficient, and somewhat unexpected. I'm not sure what the right answer is. 

Of course, there's another totally different way to handle this: I suspect it might be possible to use the variables themselves as type parameters and move operations like `subsmap()` to compile-time via clever metaprogramming. That would be similar to the way [AxisArrays.jl](https://github.com/JuliaArrays/AxisArrays.jl) allows type-stable axis lookup. That's a *huge* change, but it might be fun to explore anyway. I'll keep thinking about it.

**Comment by blegat** [comment]
Yes, I have already thought about using variables as parameter for the type. It would allow to differentiate between
```
2xy + 2xy -> Term
2xy + 2x^2 -> Polynomial
```
At the moment, the two gives `Polynomial`s.
However, it is good not to have too many different types, e.g. when you want to store elements into an array, so I feel like it is ok if things get the Polynomial type "early".

As for `subs`. I would tend to say:
* `p(y, x)` replace all the variables by `y` using the ordering `x` -> Returns the scalar type
* `subs(p, y, x)` replace some of the variables, i.e. those in `x`, by `y` -> Returns a `Polynomial`

**Comment by blegat** [comment]
I see you are trying this out [here](https://github.com/rdeits/StaticPolynomials.jl). I would be interested to see the benchmarks differences !
If it works, we could define abstract types such as `AbstractPolynomials`, ... to share some functions and make `StaticPolynomials` usable for `PolyJuMP`. Of course, if `StaticPolynomials` is faster 100% of the times, we can just replace the current implementation with it ^^

**Comment by rdeits** [comment]
OK, I like the distinction between call overloading and subs. That sounds good.

And yes, I started trying out the static polynomial idea. It's a very incomplete experiment so far, but the performance is a whole lot better. For example, given variables `x` and `y`, it is about 10,000 times faster to compute `x + y` with that package than with MultivariatePolynomials. 

There's also another implementation I'm considering it n which we use StaticArrays for all the vectors (and thus make the number of variables a type parameter) but don't actually put the variables themselves into the parameters. That would probably be less efficient but might be more flexible.

In any case, I totally agree that either making that package compatible with this one via some common abstracts would be great, and using what I learn there to improve the internals here would be even better.

**Comment by blegat** [comment]
Yes, using `immutable` types and static vectors such as `SVector` or `NTuple` instead of `Vector` would be an improvement. I do not consider the polynomials as mutable anyway

**Comment by rdeits** [comment]
I realized this morning that there may be an issue with immutable Polynomial types: it makes polynomial arithmetic and simplification type-unstable. That's because adding a new term to a polynomial will either add a new term or increase the coefficient of an existing term, but the resulting Polynomial type will be different in each case. Likewise, simplification might delete a term with a zero coefficient, so the resulting type will depend on the value of the coefficients, not just the Term types. 

The first issue can be resolved by also making the exponents parameters of the Term. That would play nicely with the new changes in v0.6 to lower `x^n` as `x^Val{n}` for literal n. It does, however, create a type-instability when raising a variable to a non-literal exponent, and it doesn't resolve simplification. But it still might be an improvement. 

One final option is to keep Term and Monomial static, but use a plain `Vector` to store the terms in a polynomial. That's more expensive, but it makes arithmetic and simplification type-stable. Hard to say which of these is best.

**Comment by rdeits** [comment]
One more fun fact: making the exponents part of the Monomial type parameters would probably let us use Horner's method to do polynomial evaluation, which could be another performance and accuracy improvement.
